File: 백준/Gold/9663. N－Queen/N－Queen.py
N = int(input())

def solve():
    board = [[0 for _ in range(N)] for _ in range(N)]
    res = 0
    leftDiag = set()
    col = set()
    rightDiag = set()

    def backtrack(r):
        nonlocal res
        if r == N:
            res += 1

            return

        for c in range(N):
            left, right = r + c, r - c
            if left in leftDiag or right in rightDiag or c in col:
                continue
            board[r][c] = 1
            leftDiag.add(left)
            col.add(c)
            rightDiag.add(right)
            backtrack(r + 1)
            leftDiag.remove(left)
            col.remove(c)
            rightDiag.remove(right)
            board[r][c] = 0

    backtrack(0)

    print(res)

solve()

File: 백준/Gold/9663. N－Queen/N－Queen.swift
import Foundation

var N = Int(readLine()!)!

func solve(_ N: Int) {
    var res = 0
    var col: [Bool] = Array(repeating: false, count: N)
    var leftDiag: [Bool] = Array(repeating: false, count: 2 * N - 1)
    var rightDiag: [Bool] = Array(repeating: false, count: 2 * N - 1)
    
    func backtrack(_ r: Int) {
        if r == N {
            res += 1
            
            return
        }
        
        for c in 0..<N {
            var left = r + c, right = c - r + (N - 1)
            if leftDiag[left] || rightDiag[right] || col[c] {
                continue
            }
            leftDiag[left] = true
            rightDiag[right] = true
            col[c] = true
            backtrack(r + 1)
            leftDiag[left] = false
            rightDiag[right] = false
            col[c] = false
        }
        
        return
    }
    
    backtrack(0)
    
    print(res)
    
    return
}

solve(N)